examples/rvc-app/rvc-common/src/rvc-device.cpp

```diff
@@ -0,0 +1,107 @@
+#include "rvc-device.h"
+
+using namespace chip::app::Clusters;
+
+void RvcDevice::Init()
+{
+    mRunModeInstance.Init();
+    mCleanModeInstance.Init();
+    mOperationalStateInstance.Init();
+}
+
+void RvcDevice::HandleRvcRunChangeToMode(uint8_t NewMode, ModeBase::Commands::ChangeToModeResponse::Type & response)
+{
+    uint8_t currentMode = mRunModeInstance.GetCurrentMode();
+
+    switch (NewMode)
+    {
+    case RvcRunMode::ModeMapping: {
+        // change to mapping only allowed from the Idle state.
+        if (currentMode != RvcRunMode::ModeIdle)
+        {
+            response.status = to_underlying(ModeBase::StatusCode::kGenericFailure);
+            response.statusText.SetValue(chip::CharSpan::fromCharString("Change to the mapping mode is only allowed from idle"));
+            return;
+        }
+
+        mRunModeInstance.UpdateCurrentMode(NewMode);
```

Comment:
Before changing run modes to anything other than Idle, should we also check if the operational state is Error?

Reply:
In this particular example we never call `OnOperationalErrorDetected`, so the operational state can never be in error. However, the `UpdateCurrentMode` would reset the error for us in this case. I think that the logic that you are suggesting may well exist on an actual robot but for this example I don't think it's necessary.